python/lsst/daf/butler/registries/sqlRegistry.py

```diff
@@ -954,3 +954,83 @@ def queryDatasets(self, datasetType: DatasetTypeExpression, *,
                     yield ref
             else:
                 yield from bestRefs.values()
+
+    def _insert(self, table, values, onConflict=None, retryLimit=3):
+        """Insert new records into a table, with conflict resolution options.
+
+        Parameters
+        ----------
+        table : `sqlalchemy.Table`
+            Table to insert new records into.
+        values : `list` [`dict`]
+            Sequence of dictionaries with values for new records.
+        onConflict: `str`, optional
+            Option for conflict resolution, can be one of "ignore" or
+            "replace". By default no conflict resolition is performed
+            and conflicts will cause immediate exceptions.
+        retryLimit : `int`, optional
+            Number of retries for insertion.
+
+        Note
+        ----
+        Conflict resolution is based on table primary key only, if there are
+        other unique constraints defined for a table they are not checked and
+        can result in `IntegrityError` exceptions.
+
+        Even with conflict resolution options it is possible that inserts
+        will generate conflicts due to concurrency and implementation details
+        of transaction isolation. When it happens the only reasonable course
+        of action is to restart transaction and repeat the whole operation.
+        Conflicts can also appear due to violation of other non-PK constraints
+        and it is not possible to distinguish those. To avoid infinite looping
+        on non-PK constraint violations this method only performs few retries.
+
+        This method needs to handle transactions itself, do not call it if you
+        are already in a transaction.
```

Comment:
This is a tricky part - the reason why it has to be done in a completely separate transaction is that in some cases (with concurrent updates and depending on backend) integrity errors could only be detected on COMMIT. Nested transactions do not do COMMIT, only top-level transactions do. So if we allow this transaction to be nested into another one then we lose logic for catching COMMIT-time errors. This also applies to other cases in `SqlRegistry` where we catch `IntegrityError` inside `@transactional` methods - depending on circumstances this exception could happen in the decorator itself and not in the try/except code.

Reply:
Gotcha.  But in that case I think we need to move the begin/commit/retry logic out of this function (and probably into a context manager), even if that means the scope of what we need to retry increases - we do need higher-level transactions so logical operations that involve multiple tables can't in the middle and leave things in an inconsistent state.

It might be possible to move the retry logic into a decorator, but we'd have to be very careful with how we used it.  This will be tricky regardless, and require a lot more care.  But I'm hoping that if we can define our operations such that operations that can recover from a conflict always use special ignore or replace syntax (and don't care which of any concurrent writes "win" in either case).  In that case, we shouldn't get failures on COMMIT anyway (right?), and of course operations that can't recover from conflicts don't need to retry.  Does that make sense?

Reply:
Even with special conflict resolution syntax it could be possible that COMMIT generates an error in case of concurrent updates - this is all very much implementation-dependent and can also vary with different transaction isolation policies.
I think we indeed can move retry logic into a reusable method, I'd probably prefer context manager over decorator to have more explicit control of what needs to be retried and what recovery options should be taken.